compiler/src/iree/compiler/GlobalOptimization/UninitializedValueValidation.cpp

```diff
@@ -0,0 +1,60 @@
+// Copyright 2025 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/GlobalOptimization/Passes.h"
+#include "llvm/ADT/SmallSet.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir::iree_compiler::GlobalOptimization {
+
+#define GEN_PASS_DEF_UNINITIALIZEDVALUEVALIDATIONPASS
+#include "iree/compiler/GlobalOptimization/Passes.h.inc"
+
+namespace {
+
+class UninitializedValueValidationPass final
+    : public impl::UninitializedValueValidationPassBase<
+          UninitializedValueValidationPass> {
+public:
+  void runOnOperation() override {
+    Operation *op = getOperation();
+
+    WalkResult walkResult = op->walk([&](linalg::LinalgOp linalgOp)
+                                         -> WalkResult {
+      Block &body = linalgOp->getRegion(0).front();
+      llvm::SmallSet<int, 4> consumedOperands;
+      for (Operation &op : body.getOperations()) {
+        for (auto operand : op.getOperands()) {
+          for (auto [i, arg] : llvm::enumerate((body.getArguments()))) {
+            if (operand == arg) {
+              consumedOperands.insert(i);
+              break;
+            }
+          }
+        }
+      }
+      for (int i : consumedOperands) {
+        Operation *defOp = linalgOp->getOperands()[i].getDefiningOp();
+        if (isa_and_nonnull<tensor::EmptyOp>(defOp)) {
+          linalgOp.emitOpError(
+              "has an uninitialized operand (produced by a tensor.empty op)");
```

Comment:
`tensor.empty` is totally fine to have as an input operand. It's only a problem if the data from the empty is used.

Reply:
that's what I thought too - it likely needs to be something on the linalg verifier to check the contents of the op and determine if there are loads - if that's not prohibitively expensive

Reply:
Makes sense - that's what I thought too - local analysis is also only sometimes going to trigger this (if the empty comes from a select, a block/func arg, a global load, etc) it won't catch it, so since it's not an error we can count on being emitted and not strictly an error we want to generate I'm not sure there's much we could do here.

Reply:
done. (note: the loop here is not over all operands but only over operands that are actually used by the block).